### PR TITLE
Ticket #1676: Require investigator for approval

### LINKED ISSUE #1676: Require investigator to be assigned before domain application can be approved
### Issue description

Currently, an application can be approved without requiring the investigator be assigned. the investigator be identified so we know the analyst who is worked the application.  To ensure accountability, the system should require the investigator be provided as the application move through the states.  

### Acceptance criteria

- [ ] the investigator field must be defined before an application can move to  from the initial state (i.e. "In review", "Approved").

### Additional context

Identified in pre-launch testing. See [issues](https://docs.google.com/document/d/1GvNzvdFOHL8jIiuDfmcxAFXeXXpbkoQx4XkRtAfSVOw/edit)

### Links to other issues

_No response_

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #1676 

## Changes

- admin.py
  - Modified __init__ for DomainApplicationAdminform to swap between what field transitions it displays depending on if investigator is none or not
  - Modified clean such that it injects form level errors when investigator is invalid
  - Small refactor of save_model. **Fixes a bug where new DomainApplication could not be added through the dashboard**
 - fixtures_applications.py
   - Added `transition.atomic()` to many of the functions here. **This will improve load times of app startup (+ running fixtures in general) while behaving the same as .save() did before**
   - Modified _approve_applications to first make sure an investigator is assigned before attempting to approve it

## Context for reviewers
This PR requires that an investigator be assigned when making the following transitions: "In review", "Action Needed", "Approved", "Rejected", "Ineligible". When going to a different status, no error should be thrown (such as submitted).

One thing to note is that the vast majority of this PR by code count is unit tests because of how our current setup is defined. I've done a bit of refactoring to them but I'm surer there still more work to do.

## Setup
1. Go to any domain application in django admin that does not have an investigator assigned. 
2. Making any edit for "Status" ONLY should be disallowed (you can edit everything else - notes, addresses, etc)
3. Try changing the status to one of the following:  "In review", "Action Needed", "Approved", "Rejected", "Ineligible". 
4. A form error should be thrown. Add a status, and continue.
5. Edits should now be allowed

This PR also includes the following fixes:
- A small bug fix for the DomainApplication create button in django admin
- Improves load times of fixtures

To test the create button fix:
1. Go to create a new domain application _within the dja dashboard_ (green add button)
2. Create it. No error should occur

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/admin.py [hunk 1]
```diff
@@ -1054,74 +1127,24 @@ def custom_election_board(self, obj):
...
-                            # status to what it was before the admin user changed it and
-                            # let the fsm method set it.
-                            obj.status = original_obj.status
-                            selected_method()
-
-                    super().save_model(request, obj, form, change)
-        else:
+        # If obj is none, something went very wrong.
+        # The form should have blocked this, so lets forbid it.
+        if not obj:
```
Comment (CocoByte) [inline src/registrar/admin.py:1136]: Thorough error-checking as always.  Thanks for the code comments :)

File: src/registrar/fixtures_applications.py [hunk 2]
```diff
@@ -184,6 +184,14 @@ def load(cls):
             logger.warning(e)
             return
 
+        # Lumped under .atomic to ensure we don't make redundant DB calls.
+        # This bundles them all together, and then saves it in a single call.
+        with transaction.atomic():
```
Comment (CocoByte) [inline src/registrar/fixtures_applications.py:189]: I did not know about this.  Thanks for the learning moment.

Comment (CocoByte) [reply src/registrar/fixtures_applications.py:189]: You should highlight this fragment to the team in demo (IMO)

File: src/registrar/utility/errors.py [hunk 3]
```diff
@@ -71,6 +71,52 @@ def get_error_message(self, code=None):
...
+
+    APPROVE_DOMAIN_IN_USE = 1
+    NO_INVESTIGATOR = 2
+    INVESTIGATOR_NOT_STAFF = 3
+    INVESTIGATOR_NOT_SUBMITTER = 4
+
+
+# (Q for reviewers) What should this be called?
+# Not a fan of this name.
+class ApplicationStatusError(Exception):
```
Comment (CocoByte) [inline src/registrar/utility/errors.py:93]: Perhaps "FSMErrorMessages"

File: src/registrar/utility/errors.py [hunk 4]
```diff
@@ -71,6 +71,52 @@ def get_error_message(self, code=None):
         return self._error_mapping.get(code)
 
 
+# (Q for reviewers) What should this be called?
+# Not a fan of this name.
+class FSMErrorCodes(IntEnum):
+    """Used when doing FSM transitions.
```
Comment (CocoByte) [inline src/registrar/utility/errors.py:77]: I like this name

**Review by CocoByte — COMMENTED** [review]
Overall, nice work on the code!  I answered one of your "questions for reviewers".  And I think your test coverage is good.  I wouldn't add any more tests.

This will need a design review, though.  Especially for the wording of the error message.  My guess is we will want something more descriptive

**Review by Katherine-Osos — CHANGES_REQUESTED** [review]
- Can you update the error message to read: Investigator is required for this status
- The AC says that users can remove an Investigator when moving a request back to the “submitted” status. But when I select that status, there isn’t a way to select “none” for investigator. So I'm still forced to have an investigator assigned. Need a way to select none/null in that use case.

**Review by PaulKuykendall — APPROVED** [review]
I tested all the use cases in the acceptance criteria, and they all passed. One comment in this PR is that the request should become read only if there is not an investigator assigned. However, that is not one of the acceptance criteria nor part of this ticket. So the fact that the request is editable without an investigator assigned, is the intended behavior.

From the product perspective, this work meets all of the criteria from the initial ticket.

For further detail on my testing results, [see this Slack thread.](https://cisa-corp.slack.com/archives/C06M4P86V47/p1709780411524529)

**Review by therealslimhsiehdy — COMMENTED** [review]
LGTM to me so far -- have a few questions for contextual reasons. Hefty amount of tests, but appreciate the thoroughness!

File: src/registrar/assets/sass/_theme/_admin.scss [hunk 5]
```diff
@@ -275,11 +275,6 @@ h1, h2, h3,
     }
 }
 
-// Hides the "clear" button on autocomplete, as we already have one to use
```
Comment (therealslimhsiehdy) [inline src/registrar/assets/sass/_theme/_admin.scss:278]: **Question:** Why was this removed?

File: src/registrar/admin.py [hunk 6]
```diff
@@ -105,6 +111,73 @@ def __init__(self, *args, **kwargs):
...
+
+        is_valid = False
+
+        # Check if an investigator is assigned. No approval is possible without one.
+        error_message = None
+        if investigator is None:
+            # Lets grab the error message from a common location
+            error_message = FSMApplicationError.get_error_message(FSMErrorCodes.NO_INVESTIGATOR)
+        elif not investigator.is_staff:
+            error_message = FSMApplicationError.get_error_message(FSMErrorCodes.INVESTIGATOR_NOT_STAFF)
```
[the remark this answers is not in the document]
Comment (therealslimhsiehdy) [reply src/registrar/admin.py:172]: Phew, thank you for the clarification! You also have tests for this, which is a great in case as well.

**Review by therealslimhsiehdy — APPROVED** [review]
Screenshot 2024-03-07 at 11 10 23 AM

Confirmed the AC with Zander that for both Admin and Analyst the STATUS for a Domain Application you cannot change without an investigator, but you can still edit all other fields. Tested this on Zanders sandbox. 

Investigator code change: Looks good, along with the tests which were very thorough for the amount of cases there are with staff, nonstaff if that case were to ever arise somehow, different transition and states, etc and error checking as well. 

Refactor: Looks good.

The `transition.atomic()` was also a nice touch as well, and def recommend demoing. 

Heavy PR, but great work!

File: src/registrar/admin.py [hunk 7]
```diff
@@ -1012,6 +1028,71 @@ def save_model(self, request, obj, form, change):
                 "This action is not permitted for applications with a restricted creator.",
             )
 
+            return None
+        
+        if change:
```
[the remark this answers is not in the document]
Comment (abroddrick) [reply src/registrar/admin.py]: reminder: if you find a bug make a ticket for it so we can be sure it's linked as part of the resolve. Also would suggest with a PR this big a bug fix would best go into another PR instead. No changes needed though on this one since I see it already has two approvals. Just more of an fyi

File: src/registrar/admin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (abroddrick) [reply src/registrar/admin.py:172]: **minor suggestion** If it can never happen, which I agree, that's a good thing to have a comment by
